sanitycheck: add option to set gcov tool

nashif opened the pull request:
When building with older versions of the toolchain we need to use gcov from
the same tool set, otherwise data formar might not be compatible.

Added an option to point to the gcov tool from the same toolchain used to
build the code. By default the host gcov is used.

Relates to #12571

nashif committed:
sanitycheck: link with --coverage only for native_posix

Link with --coverage only when building with the host toolchain.

Relates to #12571

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
sanitycheck: add option to point to gcov-tool

When building with older versions of the toolchain we need to use gcov
from the same tool set, otherwise data formar might not be compatible.

Added an option to point to the gcov tool from the same toolchain used
to build the code. By default the host gcov is used.

Relates to #12571

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

AdithyaBaglody commented on scripts/sanitycheck:
```diff
@@ -2675,6 +2677,8 @@ def parse_arguments():
         which will ultimately disable ccache.
         """
     )
+    parser.add_argument("--gcov-tool", default="gcov",
```
since this script is specific to zephyr, can we take the default gcov binary from the SDK?.
We don't need to worry about host tools.

nashif replied:
no we can't, the Zephyr SDK is not the only way we build Zephyr, we also need to support the fact the native_posix is built using the host toolchain, and not the SDK.
Sanitycheck itself also does not know anything about the toolchain, To make all of this work smoothly, the gcov data needs to be generated by the application cmakefile which then would used whatever comes with the toolchain, regardless if it is the SDK or the host toolchain, but this is a story for another time.

AdithyaBaglody replied:
oh k. my only concern was that of the shippable.

AdithyaBaglody reviewed (dismissed):
Build issue is resolved but the sanitycheck is not generating reports yet. 
GCOV reports with just CONFIG_COVERAGE=y in sample/hello_world is working but via sanitycheck no result or output seen.

cherry pick #12723  to test run on mps2_an385

nashif commented:
> Build issue is resolved but the sanitycheck is not generating reports yet.
> GCOV reports with just CONFIG_COVERAGE=y in sample/hello_world is working but via sanitycheck no result or output seen.

this works just fine with other tests and samples, not sure what is going on with the hello_world.

nashif commented:
> not sure what is going on with the hello_world.

this has nothing to do with this change, qemu just exits immediately and no gcov output is captured. try with some other samples/tests, it should work fine.

nashif committed:
scripts: gen_gcov_files: report if data capture is not complete

If we get incomplete data, report on the screen.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

nashif committed:
sanitycheck: Handle data over UART from gcov

Make sure we capture data from gcov and do not timeout before all the
data has been captured. Also report on incomplete data capture.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>

AdithyaBaglody commented:
> this works just fine with other tests and samples,

are you able to see the coverage reports?because I am not able to see them. 
Please correct me if i am wrong. i am running the following command 
`sanitycheck --coverage --coverage-platform mps2_an385 -p mps2_an385 -T tests/kernel/semaphore --gcov-tool $gcov_new`

nashif commented:
> are you able to see the coverage reports?because I am not able to see them.
> Please correct me if i am wrong. i am running the following command
> `sanitycheck --coverage --coverage-platform mps2_an385 -p mps2_an385 -T tests/kernel/semaphore --gcov-tool $gcov_new`

that is correct. Try with the updated version of this PR.

AdithyaBaglody commented:
Still the same issue persists. I am guessing LCOV is doing something weird. 
`geninfo: WARNING: /home/adithyan/zephyr/zephyr1/sanity-out/mps2_an385/samples/hello_world/sample.helloworld/CMakeFiles/app.dir/src/main.c.gcno: Overlong record at end of file!`

nashif commented:
```
i9:zephyr(gcov_fix): sanitycheck --coverage --coverage-platform mps2_an385 -p mps2_an385  --gcov-tool /home/nashif/Work/SDK/0.9.5/sysroots/x86_64-pokysdk-linux/usr/bin/i586-zephyr-elf/i586-zephyr-elf-gcov -T samples/hello_world/
JOBS: 40
Cleaning output directory /home/nashif/Work/github/zephyr/sanity-out
Building testcase defconfigs...
1 tests selected, 156 tests discarded due to filters
total complete:    1/   1  100%  failed:    0
Generating coverage files...
HTML report generated: /home/nashif/Work/github/zephyr/sanity-out/coverage/index.html
1 of 1 tests passed with 0 warnings in 15 seconds`
```

AdithyaBaglody commented:
it seems to be a bug in lcov. https://bugzilla.redhat.com/show_bug.cgi?id=1552042
The solution is to use the 1.13 version of lcov.
it works once on the new version.

nashif committed:
samples: mpu_stack_guard_test: adapt filters

Keep filters the same for both tests, sample does not work on x86.

Signed-off-by: Anas Nashif <anas.nashif@intel.com>
